lib/facter/util/xendomains.rb

```diff
@@ -2,6 +2,10 @@
 #
 module Facter::Util::Xendomains
   XEN_COMMANDS = ['/usr/sbin/xl', '/usr/sbin/xm']
+  if File.file?('/usr/lib/xen-common/bin/xen-toolstack')
```

Comment:
Putting this logic inside of the `self.xen_commands` method would be preferable, and make this much easier to test. It could look like this:

    def self.xen_command
      if File.file?('/usr/lib/xen-common/bin/xen-toolstack')
        xen_toolstack_cmd = Facter::Util::Resolution.exec('/usr/lib/xen-common/bin/xen-toolstack')
        if xen_toolstack_cmd
          xen_toolstack_cmd.chomp
        else
          nil
        end
      else
        XEN_COMMANDS.find { |cmd| Facter::Util::Resolution.which(cmd) }
      end
    end

Note that we don't need to alter the constant `XEN_COMMANDS` array if we know that `xen-toolstack` is available. In my understanding, `xen-toolstack` will return the full path of the selected tool, either `/usr/sbin/xl` or `/usr/sbin/xm` (or `nil`, which is accounted for here), so all we need to do is return its output.

No need to run `which` on the returned path from `xen-toolstack`, as it's already an absolute path according to your specs. Correct me if I'm wrong on this, and if so we'd just need to return `Facter::Util::Resolution.which(xen_toolstack_cmd.chomp)` instead.